### DISCUSSION TITLE
How do I create a middleware that can read cookie

### DISCUSSION THREAD

**Discussion by finlaydotb** [actix/actix-web#3024] [Q&A]
I am setting data in a httpOnly cookie, this means that the javascript cannot directly interact with it, but on each request it is sent to the backend.

The content of the cookie is used for authorization purposes. Now I can include logic in every handler to extract the content of the cookie but I thought that will be tedious.  

So I was thinking to extract the logic to get the cookie value and perform the authorization logic in a middleware. But unfortunately I have not been able to find out.

The closes I found is based off using `actix_session` but that does not fit my bill because it will still involve injecting `Session` into every handler, and then perform the authorization checks.  I'll the handlers don't contain this logic and hence why i will love to put it in a middleware.

Any ideas on how to go about this?

**Comment by Samielakkad** [comment]
Yes, middleware is a good place for this if you want authorization to happen before handlers run.

At the simplest level, a middleware can read an HttpOnly cookie from the incoming `ServiceRequest` like this:

```rust
fn call(&self, req: ServiceRequest) -> Self::Future {
    if let Some(cookie) = req.cookie("auth_token") {
        let token = cookie.value();
        // validate token here
    }

    let fut = self.service.call(req);
    Box::pin(async move { fut.await })
}
```

If validation succeeds, a useful pattern is to insert the authenticated user/claims into request extensions so handlers can access the already-validated identity without parsing the cookie again:

```rust
use actix_web::{
    dev::{forward_ready, Service, ServiceRequest, ServiceResponse, Transform},
    Error, HttpMessage, HttpResponse,
};
use futures_util::future::{ready, LocalBoxFuture, Ready};

#[derive(Clone, Debug)]
pub struct AuthUser {
    pub user_id: String,
}

pub struct AuthMiddleware;

impl<S, B> Transform<S, ServiceRequest> for AuthMiddleware
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type InitError = ();
    type Transform = AuthMiddlewareInner<S>;
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ready(Ok(AuthMiddlewareInner { service }))
    }
}

pub struct AuthMiddlewareInner<S> {
    service: S,
}

impl<S, B> Service<ServiceRequest> for AuthMiddlewareInner<S>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;

    forward_ready!(service);

    fn call(&self, mut req: ServiceRequest) -> Self::Future {
        let Some(cookie) = req.cookie("auth_token") else {
            return Box::pin(async move {
                Err(actix_web::error::ErrorUnauthorized("missing auth cookie"))
            });
        };

        // Replace this with your real validation/JWT/session lookup.
        let user = AuthUser {
            user_id: cookie.value().to_owned(),
        };

        req.extensions_mut().insert(user);

        let fut = self.service.call(req);
        Box::pin(async move { fut.await })
    }
}
```

Then in handlers that need the identity:

```rust
async fn private_handler(req: HttpRequest) -> HttpResponse {
    let user = req.extensions().get::<AuthUser>().cloned();
    // use user
    HttpResponse::Ok().finish()
}
```

Two practical notes:

- Put the middleware only around protected scopes/routes. Public routes like login/register should usually be outside that wrapped scope.
- HttpOnly only prevents JavaScript from reading the cookie; the browser will still send it to your backend automatically when domain/path/SameSite/Secure rules allow it. That is exactly what makes it useful for this pattern.

**Reply by robjtede** [reply]
prefer `middleware::from_fn` these days
